pkg/controllers/user/ingress/ingress.go

```diff
@@ -47,129 +43,119 @@ func (c *Controller) sync(key string, obj *v1beta1.Ingress) error {
 		return nil
 	}
 
-	serviceToPort := make(map[string]string)
-	serviceToKey := make(map[string]string)
+	expectedServices, err := generateExpectedServices(state, obj)
+	if err != nil {
+		return err
+	}
+
+	existingServices, err := getIngressRelatedServices(c.serviceLister, obj, expectedServices)
+	if err != nil {
+		return err
+	}
+
+	// 1. clean up first, delete or update the service is existing
+	for _, service := range existingServices {
+		shouldDelete, toUpdate := updateOrDelete(obj, service, expectedServices, c.needNodePort)
+		if shouldDelete {
+			if err := c.services.DeleteNamespaced(obj.Namespace, service.Name, &metav1.DeleteOptions{}); err != nil {
+				return err
+			}
+			continue
+		}
+		if toUpdate != nil {
+			if _, err := c.services.Update(toUpdate); err != nil {
+				return err
+			}
+		}
+		// don't create the services which alread exist
+		delete(expectedServices, service.Name)
+	}
+
+	// 2. create the new services
+	for _, ingressService := range expectedServices {
+		toCreate := ingressService.generateNewService(obj)
```

Comment:
lets pass servicetype as an argument to the generateNewService (ClusterIP vs NodePort) call. It will be cleaner this way.